### PR TITLE
[JUJU-113] Refactor upgrade controller command

### PR DESCRIPTION
**Pull request by ycliuhw**
The `upgrade-controller` command currently fetches available agent versions for upgrading on the client-side which makes the command itself complicated and difficult to manage consistencies. Therefore, we moved the agent versions (`jujud-operator` image tags) fetching for CAAS to the server-side (This also causes the `Client` facade upgraded to version `4`). 

Driveby fix: currently we compare the available versions with `version.Current` from CLI which is not right because we are upgrading controller rather than CLI. Now we get the current version from the model config inside the facade method to compare with.

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages~
 - [x] Comments answer the question of why design decisions were made

## QA steps

```console
$ juju bootstrap microk8s k1 --config caas-image-repo="'$(cat ghcr-cred.json)'"

$ juju add-model t1 microk8s

$ juju status --relations --color --storage -m k1:t1 --format yaml
model:
  name: t1
  type: caas
  controller: k1
  cloud: microk8s
  region: localhost
  version: 2.9.19
  model-status:
    current: available
    since: 15 Nov 2021 15:22:43+11:00
  sla: unsupported
machines: {}
applications: {}
storage: {}
controller:
  timestamp: 15:43:24+11:00
Model "t1" is empty.

$ juju upgrade-model -m controller --agent-stream=develop
best version:
    2.9.20
started upgrade to 2.9.20

$ juju upgrade-model -m t1 --agent-stream=develop
best version:
    2.9.20
started upgrade to 2.9.20


$ juju bootstrap localhost k1 --config caas-image-repo="'$(cat ghcr-cred.json)'"

$ juju add-model t1 localhost

$ juju upgrade-model -m controller --agent-stream=develop --debug

$ juju upgrade-model -m t1 --agent-stream=develop --debug

$ juju bootstrap aws/ap-southeast-2 k1 --config caas-image-repo="'$(cat ghcr-cred.json)'"

$ juju add-model t1 aws/ap-southeast

$ juju upgrade-model -m t1 --agent-stream=develop
best version:
    2.9.19.5
started upgrade to 2.9.19.5

$ juju add-k8s --gke gke --debug

$ juju add-model t2 gke

$ juju upgrade-model -m t2 --agent-stream=develop --dry-run
best version:
    2.9.20
upgrade to this version by running
    juju upgrade-model

```

## Documentation changes

No UX changes;

## Bug reference

https://bugs.launchpad.net/juju/+bug/1944801

### CODE REVIEW

File: apiserver/facades/client/client/client.go [hunk 1]
```diff
@@ -154,14 +161,24 @@ func NewFacadeV1(ctx facade.Context) (*ClientV1, error) {
...
+// NewFacadeV3 creates a version 3 Client facade to handle API requests.
+func NewFacadeV3(ctx facade.Context) (*ClientV3, error) {
+	client, err := NewFacade(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return &ClientV3{client}, nil
+}
+
+// NewFacade creates a version 4 Client facade to handle API requests.
```
Comment (hpidcock) [inline apiserver/facades/client/client/client.go:180]: would be good to mention what v4 adds

File: apiserver/facades/client/client/client.go [hunk 2]
```diff
@@ -792,14 +812,127 @@ func (c *Client) AbortCurrentUpgrade() error {
 }
 
 // FindTools returns a List containing all tools matching the given parameters.
-func (c *Client) FindTools(args params.FindToolsParams) (params.FindToolsResult, error) {
+func (c *ClientV3) FindTools(args params.FindToolsParams) (params.FindToolsResult, error) {
+	logger.Criticalf("ClientV3.FindTools args %#v", args)
```
Comment (hpidcock) [inline apiserver/facades/client/client/client.go:816]: obviously remove this before landing or make it a more useful trace

File: apiserver/facades/client/client/client.go [hunk 3]
```diff
@@ -792,14 +812,127 @@ func (c *Client) AbortCurrentUpgrade() error {
...
+func (c *Client) FindTools(args params.FindToolsParams) (result params.FindToolsResult, err error) {
+	defer func() {
+		logger.Criticalf("Client.FindTools args -> %#v, params.FindToolsResult -> %#v, error -> %#v", args, result, err)
+	}()
+
+	if err := c.checkCanWrite(); err != nil {
+		return params.FindToolsResult{}, err
+	}
+	result, err = c.api.toolsFinder.FindTools(args)
+	if err == nil {
```
Comment (hpidcock) [inline apiserver/facades/client/client/client.go:833]: combine these two expressions into
`err == nil && result.Error != nil`

File: apiserver/facades/client/client/client.go [hunk 4]
```diff
@@ -792,14 +812,127 @@ func (c *Client) AbortCurrentUpgrade() error {
...
+		if result.Error != nil {
+			err = result.Error
+			// We need to deal with older controllers.
+			if strings.HasSuffix(result.Error.Message, "not valid") {
```
Comment (hpidcock) [inline apiserver/facades/client/client/client.go:837]: I thought there was a params function for unpacking errors, if there is, we should use that here

File: apiserver/facades/client/client/client.go [hunk 5]
```diff
@@ -792,14 +812,127 @@ func (c *Client) AbortCurrentUpgrade() error {
...
+				// TODO: probably we can remove this!!
+				err = errors.NotValidf("finding stream data for this model")
+			}
+			if params.IsCodeNotFound(err) {
+				err = errors.NewNotFound(err, "finding stream data for this model")
+			}
+		}
+	}
+	if err != nil && !errors.IsNotFound(err) {
+		return result, errors.Trace(err)
```
Comment (hpidcock) [inline apiserver/facades/client/client/client.go:847]: would be good to annotate this

File: apiserver/facades/client/client/client.go [hunk 6]
```diff
@@ -792,14 +812,127 @@ func (c *Client) AbortCurrentUpgrade() error {
...
+	if imageRepoDetails.Empty() {
+		repoDetails, err := docker.NewImageRepoDetails(podcfg.JujudOCINamespace)
+		if err != nil {
+			return result, errors.Trace(err)
+		}
+		imageRepoDetails = *repoDetails
+	}
+	reg, err := c.registryAPIFunc(imageRepoDetails)
+	if err != nil {
+		return result, errors.Trace(err)
```
Comment (hpidcock) [inline apiserver/facades/client/client/client.go:882]: annotate here too please

File: cmd/juju/commands/synctools.go [hunk 7]
```diff
@@ -170,13 +171,14 @@ type syncToolsAPIAdapter struct {
 
 func (s syncToolsAPIAdapter) FindTools(majorVersion int, stream string) (coretools.List, error) {
 	result, err := s.syncToolsAPI.FindTools(majorVersion, -1, "", "", "")
+	if params.IsCodeNotFound(result.Error) || errors.IsNotFound(err) {
```
Comment (hpidcock) [inline cmd/juju/commands/synctools.go:174]: cmd code should not be concerned with params errors, this should be handled in the client api code

File: cmd/juju/commands/synctools.go [hunk 8]
```diff
@@ -170,13 +171,14 @@ type syncToolsAPIAdapter struct {
...
+		return nil, coretools.ErrNoMatches
+	}
 	if err != nil {
 		return nil, err
 	}
 	if result.Error != nil {
```
Comment (hpidcock) [inline cmd/juju/commands/synctools.go:180]: same here, we really shouldn't expose the underlying result params struct

File: cmd/juju/commands/upgrademodel.go [hunk 9]
```diff
@@ -375,20 +368,14 @@ func (c *upgradeJujuCommand) getControllerAPI() (ControllerAPI, error) {
 	return apicontroller.NewClient(api), nil
 }
 
+const caasStreamsTimeout = 20 * time.Second
```
Comment (hpidcock) [inline cmd/juju/commands/upgrademodel.go:371]: I assume 20 seconds is enough based on your own testing

File: cmd/juju/commands/upgrademodel.go [hunk 10]
```diff
@@ -817,10 +798,10 @@ func (c *baseUpgradeCommand) initVersions(
 	}
 	logger.Debugf("searching for %q agent binaries with major: %d", c.AgentStream, filterVersion.Major)
 	streamVersions, err := fetchStreamsVersions(client, filterVersion.Major, c.AgentStream, timeout)
-	if err != nil && !params.IsCodeNotFound(err) {
+	if err != nil && !errors.IsNotFound(err) {
```
Comment (hpidcock) [inline cmd/juju/commands/upgrademodel.go:801]: this is good

File: cmd/juju/commands/upgrademodel.go [hunk 11]
```diff
@@ -833,6 +814,43 @@ func (c *baseUpgradeCommand) initVersions(
 	}, nil
 }
 
+// fetchStreamsVersions returns simplestreams agent metadata
+// for the specified stream. timeout ensures we don't block forever.
+func fetchStreamsVersions(
+	client toolsAPI, majorVersion int, stream string, timeout time.Duration,
```
Comment (hpidcock) [inline cmd/juju/commands/upgrademodel.go:820]: kinda wish we used `context.Context` more so timeouts and cancel propagation worked more idiomatically

File: apiserver/facades/client/client/client.go (hunk 4 above)
Comment (ycliuhw) [reply apiserver/facades/client/client/client.go:837]: This code was moved here from the command level. 
I don't think we have any code dealing with not valid specific API server side error code.

File: apiserver/facades/client/client/client.go (hunk 3 above)
Comment (ycliuhw) [reply apiserver/facades/client/client/client.go:833]: Yeah, it's been changed, and this kinda error translation has been moved to API client-side.
Just not pushed yet.

File: cmd/juju/commands/upgrademodel.go (hunk 9 above)
Comment (ycliuhw) [reply cmd/juju/commands/upgrademodel.go:371]: This was moved from `upgradecontroller.go`.

File: cmd/juju/commands/upgrademodel.go (hunk 11 above)
Comment (ycliuhw) [reply cmd/juju/commands/upgrademodel.go:820]: agreed, but the facade code does not support context. So we will still have to spawn a goroutine manually when we call the facade method.

**Comment by ycliuhw** [comment]
!!build!!

**Review by hpidcock — APPROVED** [review]
LGTM

File: apiserver/facades/client/client/client.go [hunk 12]
```diff
@@ -792,14 +814,111 @@ func (c *Client) AbortCurrentUpgrade() error {
...
+	if err := c.checkCanWrite(); err != nil {
+		return params.FindToolsResult{}, err
+	}
+	model, err := c.api.stateAccessor.Model()
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	result, err = c.api.toolsFinder.FindTools(args)
+	if model.Type() != state.ModelTypeCAAS {
+		return result, errors.Annotate(err, "finding tool version from simple streams")
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply apiserver/facades/client/client/client.go:835]: We return here for non CAAS model and only do below steps for the CAAS model.
Added comments to explain it clear.

File: apiserver/facades/client/client/client.go [hunk 13]
```diff
@@ -792,14 +814,111 @@ func (c *Client) AbortCurrentUpgrade() error {
...
+	}
+	if err != nil && !errors.IsNotFound(err) || result.Error != nil && !params.IsCodeNotFound(result.Error) {
+		return result, errors.Annotate(err, "finding tool versions from simplestream")
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply apiserver/facades/client/client/client.go:838]: We convert to errors's notfound in `api` but not in `apiserver` package.

**Comment by ycliuhw** [comment]
!!build!!

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**